src/Storages/StorageAlias.cpp

```diff
@@ -0,0 +1,107 @@
+#include <Storages/StorageAlias.h>
+#include <Storages/StorageFactory.h>
+#include <Storages/checkAndGetLiteralArgument.h>
+#include <Interpreters/DatabaseCatalog.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Interpreters/Context.h>
+#include <Core/Settings.h>
+
+namespace DB
+{
+
+namespace Setting
+{
+    extern const SettingsSeconds lock_acquire_timeout;
+}
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int BAD_ARGUMENTS;
+}
+
+StorageAlias::StorageAlias(
+    const StorageID & table_id_,
+    const ColumnsDescription & columns_,
+    const ConstraintsDescription & constraints_,
+    const String & comment,
+    ContextPtr context_,
+    const StorageID & ref_table_id_)
+    : IStorage(table_id_)
+    , ref_table_id(ref_table_id_)
+{
+    if (table_id_ == ref_table_id_)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Alias table cannot refer to itself.");
+
+    StorageInMemoryMetadata storage_metadata;
+    if (columns_.empty())
+    {
+        auto ref_table = DatabaseCatalog::instance().getTable(ref_table_id, context_);
+        storage_metadata.setColumns(ref_table->getInMemoryMetadataPtr()->getColumns());
+    }
+    else
+        storage_metadata.setColumns(columns_);
```

Comment:
in which cases the columns is empty or not empty?

Reply:
I just followed the pattern of `StorageBuffer`:

https://github.com/ClickHouse/ClickHouse/blob/5b8e4cc14edd339d8cae7c4cd4cf170ffee20668/src/Storages/StorageBuffer.cpp#L166-L172

Maybe there isn't a case where `columns_` is empty.

Reply:
we should throw exception if it is not empty:
```
:) create table t (a Int32, b Int32) Engine = Log();

CREATE TABLE t
(
    `a` Int32,
    `b` Int32
)
ENGINE = Log

Query id: 63c1053e-9024-41a3-8c30-21f2046c51c5

Ok.

0 rows in set. Elapsed: 0.004 sec.

:) create table t1 (x1 String) Engine = Alias(default, t);

CREATE TABLE t1
(
    `x1` String
)
ENGINE = Alias(default, t)

Query id: 375c692a-b04c-42a2-981e-5235cefd3e9c

Ok.

0 rows in set. Elapsed: 0.003 sec.
```

Reply:
Oh, I know what you mean. I think we should throw an exception if the schema definition of alias table is not the same as the ref table's.

Reply:
StorageBuffer also has a bug:
```
:) create table t2 (x1 String) Engine = Buffer(default, t, 1, 10, 100, 10000, 1000000, 10000000, 1000000000);

CREATE TABLE t2
(
    `x1` String
)
ENGINE = Buffer(default, t, 1, 10, 100, 10000, 1000000, 10000000, 1000000000)

Query id: c2a5bb62-f71a-4527-8cbb-7dc5cb99844c

Ok.

0 rows in set. Elapsed: 0.006 sec.
```
it should throw exception as far as I understand :(

Reply:
> it should be always empty

Looks good to me.